packages/graph/hash_graph/lib/graph/src/datastore/postgres/mod.rs

```diff
@@ -25,22 +32,126 @@ impl PostgresDatabase {
+            .attach_printable_lazy(|| id.clone())
+            .attach_printable("Could not insert id")?;
+
+        Ok(())
+    }
+
+    async fn insert_data_type(
+        &self,
+        id: &Identifier,
+        data_type: &DataType,
+        created_by: AccountId,
+    ) -> Result<DataTypeRow, DatastoreError> {
+        sqlx::query_as(
+            r#"
+            INSERT INTO data_types (
+                version_id,
+                schema,
+                created_by
+            ) 
+            VALUES ($1, $2, $3)
+            RETURNING version_id, schema, created_by;
+            "#,
+        )
+        .bind(&id.version_id)
+        .bind(&data_type)
+        .bind(&created_by)
+        .fetch_one(&self.pool)
+        .await
+        .report()
+        .change_context(DatastoreError)
+        .attach_printable("Could not insert data type")
+    }
+}
+
 #[async_trait]
 impl Datastore for PostgresDatabase {
-    async fn create_data_type() -> Result<(), DatastoreError> {
-        todo!()
+    async fn create_data_type(
+        &self,
+        data_type: DataType,
+        created_by: AccountId,
+    ) -> Result<Qualified<DataType>, DatastoreError> {
+        let id = Identifier {
+            base_id: BaseId::new(Uuid::new_v4()),
+            version_id: VersionId::new(Uuid::new_v4()),
+        };
+
+        self.insert_version_id(&id).await?;
+
+        self.insert_data_type(&id, &data_type, created_by).await?;
+
+        Ok(Qualified {
+            id,
+            inner: data_type,
+            created_by,
+        })
     }
 
-    async fn get_data_type() -> Result<(), DatastoreError> {
-        todo!()
+    async fn get_data_type(&self, id: Identifier) -> Result<Qualified<DataType>, DatastoreError> {
+        let DataTypeRow {
+            schema: data_type,
+            created_by,
+            ..
+        } = sqlx::query_as(
+            r#"
+            SELECT version_id, "schema", created_by
+            FROM data_types
+            INNER JOIN ids USING (version_id)
+            WHERE version_id = $1 AND base_id = $2;
+            "#,
+        )
+        .bind(&id.version_id)
+        .bind(&id.base_id)
+        .fetch_one(&self.pool)
+        .await
+        .report()
+        .change_context(DatastoreError)
+        .attach_printable_lazy(|| id.clone())
+        .attach_printable("Could not find data type by id")?;
+
+        Ok(Qualified {
+            id,
+            inner: data_type,
+            created_by,
+        })
     }
 
-    async fn get_data_many() -> Result<(), DatastoreError> {
+    async fn get_data_type_many() -> Result<(), DatastoreError> {
         todo!()
     }
 
-    async fn update_data_type() -> Result<(), DatastoreError> {
-        todo!()
+    async fn update_data_type(
+        &self,
+        base_id: BaseId,
+        data_type: DataType,
+        updated_by: AccountId,
+    ) -> Result<Qualified<DataType>, DatastoreError> {
+        let id = Identifier {
```

Comment:
Just flagging that this sort of thing is why having clearer methods, e.g. `Identifier::new` or `Identifier::from` would be more clear here, because at first glance this looks very similar to the one used in `create_data_type` and if skimming quickly it might be easy to assume it's the same and isn't reusing the base_id.

Reply:
Addressed in https://github.com/hashintel/hash/pull/734/commits/f50585663a877a9865d5ce5ecb7a264ecbca1eae